example/js/js_image_layer.bzl

```diff
@@ -0,0 +1,260 @@
+"contains container helper functions for js_binary"
+
+load("@rules_pkg//:providers.bzl", "PackageFilegroupInfo", "PackageFilesInfo", "PackageSymlinkInfo")
+load("@rules_pkg//:pkg.bzl", "pkg_tar")
+load("@bazel_skylib//rules:write_file.bzl", "write_file")
+load("@aspect_bazel_lib//lib:paths.bzl", "to_manifest_path")
+load("@rules_python//python:defs.bzl", "py_binary")
+
+# BAZEL_BINDIR has to be set to '.' so that js_binary preserves the PWD when running inside container.
+# See https://github.com/aspect-build/rules_js/tree/dbb5af0d2a9a2bb50e4cf4a96dbc582b27567155#running-nodejs-programs
+# for why this is needed.
+_LAUNCHER_TMPL = """
+export BAZEL_BINDIR=.
+source {executable_path}
+"""
+
+def _write_laucher(ctx, executable_path):
+    "Creates a call-through shell entrypoint which sets BAZEL_BINDIR to '.' then immediately invokes the original entrypoint."
+    launcher = ctx.actions.declare_file("%s_launcher.sh" % ctx.label.package)
+
+    ctx.actions.write(
+        output = launcher,
+        content = _LAUNCHER_TMPL.format(executable_path = executable_path),
+        is_executable = True,
+    )
+
+    return launcher
+
+def _runfile_path(ctx, file, runfiles_dir):
+    return "/".join([runfiles_dir, to_manifest_path(ctx, file)])
+
+def _should_include(destination, include, exclude):
+    included = include in destination or include == ""
+    excluded = exclude in destination and exclude != ""
+    return included and not excluded
+
+def _runfiles_impl(ctx):
+    default = ctx.attr.binary[DefaultInfo]
+
+    executable = default.files_to_run.executable
+    executable_path = "/".join([ctx.attr.root, executable.short_path])
+    original_executable_path = executable_path.replace(".sh", "_.sh")
+    launcher = _write_laucher(ctx, original_executable_path)
+
+    file_map = {}
+
+    if _should_include(original_executable_path, ctx.attr.include, ctx.attr.exclude):
+        file_map[original_executable_path] = executable
+
+    if _should_include(executable_path, ctx.attr.include, ctx.attr.exclude):
+        file_map[executable_path] = launcher
+
+    manifest = default.files_to_run.runfiles_manifest
+    runfiles_dir = "/".join([ctx.attr.root, manifest.short_path.replace(manifest.basename, "")[:-1]])
+
+    files = depset(transitive = [default.files, default.default_runfiles.files])
+
+    for file in files.to_list():
+        destination = _runfile_path(ctx, file, runfiles_dir)
+        if _should_include(destination, ctx.attr.include, ctx.attr.exclude):
+            file_map[destination] = file
+
+    # executable and launcher should not go into runfiles directory so we add it to files here
+    files = depset([executable, launcher], transitive = [files])
```

Comment:
This part is confusing as `files` gets reassigned. This can easily introduce code smell as the code gets maintained over time.